src/core/mesh/qgsmeshlayerprofilegenerator.cpp

```diff
@@ -0,0 +1,157 @@
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsmeshlayerprofilegenerator.h"
+#include "qgsprofilerequest.h"
+#include "qgscurve.h"
+#include "qgsmeshlayer.h"
+#include "qgscoordinatetransform.h"
+#include "qgsgeos.h"
+#include "qgsterrainprovider.h"
+#include "qgsmeshlayerutils.h"
+
+//
+// QgsMeshLayerProfileGenerator
+//
+
+QString QgsMeshLayerProfileResults::type() const
+{
+  return QStringLiteral( "mesh" );
+}
+
+QHash<double, double> QgsMeshLayerProfileResults::distanceToHeightMap() const
+{
+  QHash<double, double> res;
+  for ( const Result &r : results )
+  {
+    res.insert( r.distance, r.height );
+  }
+  return res;
+}
+
+QgsPointSequence QgsMeshLayerProfileResults::sampledPoints() const
+{
+  return rawPoints;
+}
+
+QVector<QgsGeometry> QgsMeshLayerProfileResults::asGeometries() const
+{
+  QVector<QgsGeometry> res;
+  res.reserve( rawPoints.size() );
+  for ( const QgsPoint &point : rawPoints )
+    res.append( QgsGeometry( point.clone() ) );
+
+  return res;
+}
+
+//
+// QgsMeshLayerProfileGenerator
+//
+
+QgsMeshLayerProfileGenerator::QgsMeshLayerProfileGenerator( QgsMeshLayer *layer, const QgsProfileRequest &request )
+  : mProfileCurve( request.profileCurve() ? request.profileCurve()->clone() : nullptr )
+  , mSourceCrs( layer->crs() )
+  , mTargetCrs( request.crs() )
+  , mTransformContext( request.transformContext() )
+{
+  layer->updateTriangularMesh();
+  mTriangularMesh = *layer->triangularMesh();
+}
+
+QgsMeshLayerProfileGenerator::~QgsMeshLayerProfileGenerator() = default;
+
+bool QgsMeshLayerProfileGenerator::generateProfile()
+{
+  if ( !mProfileCurve )
+    return false;
+
+  // we need to transform the profile curve to the mesh's CRS
+  QgsGeometry transformedCurve( mProfileCurve->clone() );
+  mLayerToTargetTransform = QgsCoordinateTransform( mSourceCrs, mTargetCrs, mTransformContext );
+
+  try
+  {
+    transformedCurve.transform( mLayerToTargetTransform, Qgis::TransformDirection::Reverse );
+  }
+  catch ( QgsCsException & )
+  {
+    QgsDebugMsg( QStringLiteral( "Error transforming profile line to mesh CRS" ) );
+    return false;
+  }
+
+  mResults = std::make_unique< QgsMeshLayerProfileResults >();
+
+  // we don't currently have any method to determine line->mesh intersection points, so for now we just sample at about 100(?) points over the line
```

Comment:
I think it could be possible, but a bit tricky, to implement something that intersect lines with the mesh edges. If it is not implemented for now, I am happy there is still some room to implement it later.